### PR TITLE
Ship v1

### PR DESCRIPTION
**Pull request by huangminghuang**
This PR defines new types used by ship v1 protocols.

### CODE REVIEW

File: include/eosio/ship_protocol.hpp [hunk 1]
```diff
@@ -15,6 +15,8 @@ template <typename S>
 void to_json(const input_stream& data, S& stream) {
    return to_json_hex(data.pos, data.end - data.pos, stream);
 }
+
+constexpr const char* get_type_name(const input_stream*) { return "bytes";  }
```
Comment (heifner) [inline include/eosio/ship_protocol.hpp:19]: Should this be moved to stream.hpp ?

Comment (swatanabe-b1) [reply include/eosio/ship_protocol.hpp:19]: Probably, but there's a dependency issue: `to_json.hpp` depends on `stream.hpp`, so we can't put anything that depends on to_json in `stream.hpp` or the dependencies become circular.

File: include/eosio/ship_protocol.hpp [hunk 2]
```diff
@@ -124,19 +126,22 @@ namespace eosio { namespace ship_protocol {
 
    using request = std::variant<get_status_request_v0, get_blocks_request_v0, get_blocks_ack_request_v0>;
 
-   struct get_blocks_result_v0 {
+   struct get_blocks_result_base_v0 {
```
Comment (swatanabe-b1) [inline include/eosio/ship_protocol.hpp:129]: You can leave off the `_v0`.  This is common to `get_blocks_result_v0` and `get_blocks_result_v1`.

File: include/eosio/opaque.hpp [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+         sizeof("opaque") + std::string_view(get_type_name((T*)nullptr)).size() + sizeof("vector") + 1;
+   std::array<char, size> buffer{ 'o', 'p', 'a', 'q', 'u', 'e' };
+   (variant_type_appender{ buffer.data() + 6 } + std::string_view(get_type_name((T*)nullptr)) +
+    std::string_view("vector"));
+   buffer[buffer.size() - 1] = '\0';
+   return buffer;
+}
+
+template <typename T>
+constexpr auto opaque_type_name = get_opaque_type_name<T>();
```
Comment (swatanabe-b1) [inline include/eosio/opaque.hpp:119]: Is this actually necessary?  The type name is used for generating an ABI, and the ABI type for all opaque types is bytes.

**Pull request merged**